### ISSUE TITLE
Eval bug: Quantized K cache makes deterministic prefix-cache repeats produce different output

### ISSUE THREAD

**Issue by Redempt1onzzZZ** [ggml-org/llama.cpp#28527]
### Name and Version

llama.cpp `version: 1 (1d7ab2b)`, built with CUDA.

### Operating systems

Linux

### GGML backends

CUDA

### Hardware

NVIDIA H100 80GB HBM3.

### Models

Target: Meta-Llama-3-8B-Instruct.Q4_0.gguf.

Control: Qwen3-4B-Q4_K_M.gguf.

The target model is a local GGUF file; add the exact upstream model URL when submitting.

### Problem description & steps to reproduce

Start the server:

```bash
/root/workspace/llama.cpp/build-cuda/bin/llama-server \
  --model /root/.cache/gpt4all/Meta-Llama-3-8B-Instruct.Q4_0.gguf \
  --host 127.0.0.1 \
  --port 64400 \
  --pooling none \
  --cache-type-k iq4_nl \
  --no-kv-offload
```

Send the same deterministic long-prefix request three times. The following command uses
an approximately 9k-character prompt, which is long enough to exercise the prefix cache:

```bash
python3 - <<'PY'
import json
import subprocess

prompt = ("This is a deterministic long-context prefix for llama.cpp. " * 256).strip()
payload = {
    "model": "Meta-Llama-3-8B-Instruct.Q4_0.gguf",
    "messages": [{"role": "user", "content": prompt}],
    "temperature": 0,
    "seed": 0,
    "max_tokens": 16,
    "stream": False,
}
for i in range(3):
    out = subprocess.check_output([
        "curl", "-sS", "http://127.0.0.1:64400/v1/chat/completions",
        "-H", "Content-Type: application/json",
        "-d", json.dumps(payload),
    ], text=True)
    print(f"response_{i+1}: {out}")
PY
```

The observed target replay used the same deterministic request shape with an 8,959-character
repeated prefix. The server remained healthy and returned HTTP 200 for the requests, but the
generated text changed between the first and later prefix-cache repeats.

### Relevant log output

**Relevant log output**

```text
HTTP 200
response 1: "It seems like you're trying to test the limits of my language processing abilities by"
response 2: "It seems like you're trying to test the limits of the AI's ability to"
response 3: "It seems like you're trying to test the limits of the AI's ability to"
```

The deterministic request uses `temperature=0` and `seed=0`. Ordinary chat, streaming,
recovery, and short completion requests pass under the same server configuration. Replaying
the same configuration with Qwen3-4B-Q4_K_M.gguf passes the prefix-cache repeat check.

**Comment by sakurahello1** [comment]
Before replying I want to point out one thing. The reporter found that on an H100 with Llama-3-8B-Instruct.Q4_0.gguf, the same prefix (temperature=0, seed=0) produces different text before and after the prompt cache is reused, and concluded that KV-cache quantization causes the drift.

Decoding here is greedy. When the top two candidate logits differ by only 0.015, any numerical perturbation around 1e-2 flips the choice. The perturbation comes from floating-point accumulation order. Quantization of the model or of the KV cache is fixed across the three requests; it only determines how close the candidates are to begin with and is not what changes between runs.

I checked where the difference actually comes from and it is the compute path. The first request runs a full prefill through the matrix-matrix kernels. A cached repeat backs off one token (n_past-- under [TAG_PROMPT_LOGITS] in tools/server/server-context.cpp) and re-evaluates only that token, which goes through the matrix-vector kernels. The two accumulate in a different order, so the logits differ a little, and sometimes that flips the next token.

I reproduced it on an RTX 4060 Laptop (CUDA 12.8, master f172be7) with the same model and the exact request from the issue, adding logprobs / top_logprobs: 5 to look at the first generated token. With iq4_nl K the first request had a top-2 logprob gap of 0.015 ("I" -1.130 vs "It" -1.145). On the cached repeat the candidates moved by +4.8e-2, +1.4e-2, -1.4e-1 and +1.1e-1, the ranking flipped and the text changed; requests 2 and 3 were bit-identical to each other. With f16 K the candidates moved by the same magnitude on the repeat (+5.1e-3, -6.2e-2, +9.1e-2, +3.5e-2, -7.6e-2), but the gap there was 0.239, so nothing flipped and the text stayed the same. Running iq4_nl with cache_prompt: false, so that every request re-prefills, gave three bit-identical responses with all deltas at zero. On a CPU build with --no-kv-offload it is the other way round: iq4_nl K is bit-identical across the three requests and f16 K drifts by up to 2.4e-2, which is what you'd expect if the variable is which kernel handles the batch rather than the K-cache type.

In HPC this is generally not considered a bug; floating-point results are not expected to be batch-invariant. Anyone who needs byte-identical output across repeats can use cache_prompt: false (or --no-cache-prompt

**Comment by sakurahello1** [comment]
I think this can be closed as working as intended.
